ga4gh/datamodel/genotype_phenotype.py

```diff
@@ -0,0 +1,467 @@
+          "parentIds": [],
+          "featureSetId": "",
+          "attributes": {"vals": {}},
+          "id": annotation["id"],
+        }
+        if (not protocol.Feature.validate(d)):
+            raise exceptions.RequestValidationFailureException(d,
+                                                               protocol.
+                                                               Feature)
+
+        f = protocol.Feature.fromJsonDict(d)
+
+        (_id, _ontologySource) = self.namespace_split(
+                                       annotation[hasObject]['val'])
+
+        d = {
+            "features": [],
+            "evidence": [],
+            "environmentalContexts": [],
+            "phenotype": {
+                "id": "",
+                "type": {"ontologySource": _ontologySource,
+                         "id": _id,
+                         "name": annotation[hasObject]['label']}
+                },
+            "id": annotation['id'],
+            "description": "null"
+        }
+
+        fpa = protocol.FeaturePhenotypeAssociation.fromJsonDict(d)
+        fpa.features.append(f)
+
+        #  ECO or OBI is recommended
+        if (source in annotation):
+            if (annotation[source].__class__.__name__ != "list"):
+                annotation[source] = [annotation[source]]
+            for s in annotation[source]:
+                evidence = protocol.Evidence()
+                evidence.evidenceType = protocol.OntologyTerm()
+                (_id, _ontologySource) = self.namespace_split(s['val'])
+                evidence.evidenceType.ontologySource = _ontologySource
+                evidence.evidenceType.id = _id
+
+                evidence.evidenceType.name = ''
+                if 'label' in annotation[_evidence]:
+                    evidence.evidenceType.name = annotation[_evidence]['label']
+                fpa.evidence.append(evidence)
+                if (not protocol.Evidence.validate(evidence.toJsonDict())):
+                    raise exceptions.RequestValidationFailureException(
+                           evidence.toJsonDict(), protocol.Evidence)
+        if (not
+            protocol.FeaturePhenotypeAssociation
+                    .validate(fpa.toJsonDict())):
+            raise exceptions.RequestValidationFailureException(
+                fpa.toJsonDict(), protocol.FeaturePhenotypeAssociation)
+
+        return fpa
+
+    def namespace_split(self, _url, _separator='/'):
+        """
+        given a url return the id of the resource and the ontology source
+        """
+        from urlparse import urlparse, urlunsplit
+        o = urlparse(_url)
+        _id = o.path.split(_separator)[-1]
+        _ontologySource = urlunsplit([o[0],
+                                      o[1],
+                                      o[2].replace(_id, ''),
+                                      o[4], ''])
+        return(_id, _ontologySource)
+
+    def test_data(self, results):
```

Comment:
I don't like these two functions: test code should be in the test modules. We're complicating the production code here with tests.